op/model.py

```diff
@@ -257,6 +312,65 @@ class RelationNotFound(ModelError):
     pass
 
 
+class Status(ABC):
+    """Status values specific to applications and units."""
+
+    _statuses = {}
+
+    def __init__(self, message):
+        self.message = message
+
+    def __new__(cls, *args, **kwargs):
+        if cls is Status:
+            raise TypeError("cannot instantiate a base class")
+
+        return super().__new__(cls)
+
+    @classmethod
+    def _register_status(cls, name, type_):
+        """For use by subclasses only."""
+        cls._statuses[name] = type_
+
+    @classmethod
+    def from_string(cls, name, message):
+        return cls._statuses[name](message)
+
+    def __init_subclass__(cls):
+        super().__init_subclass__()
+        Status._register_status(cls.name, cls)
+
+class Active(Status):
+    """The unit believes it is correctly offering all the services it has been asked to offer."""
+    name = 'active'
```

Comment:
I think it makes sense to do:

def __init__(self, message=''):

for Active as well. Many charms don't have anything to report when they get to the "I'm happily running" state. Some do, but I don't think we must require it. (Certainly they could actively set it to Active(''), its more about signaling that its acceptable to not set a message)

Reply:
+1, thanks for the suggestion.

Reply:
We can make it optional in `Status` itself I think.

Reply:
My understanding was that we liked putting pressure on charmers for setting a status for something like Blocked, because you really do need to communicate to the human. And that also is reasonable for Maintenance and Waiting. So its really just Active and Unknown